### PR TITLE
Happ 445 write unit tests for coredata functionality

### PR DESCRIPTION
**Pull request by freshfawzy**
Created core data stack class for test, and wrote test cases for Patient and VaccineCard entities

### CODE REVIEW

File: BCVaccineCard/BCVaccineCard/Screens/ReusableViewControllers/GatewayFormViewController.swift [hunk 1]
```diff
@@ -555,11 +558,13 @@ extension GatewayFormViewController: HealthGatewayAPIWorkerDelegate {
         // store prefered PHN if needed here
         self.rememberedPHNSelected ? storePHNDetails() : removePHNDetailsIfNeccessary()
         hideLoader()
-        if StorageService.shared.testExists(from: result) {
+        if storageService.testExists(from: result) {
             alert(title: .duplicateTitle, message: .duplicateTestMessage)
             return
         }
-        if let id = handleTestResultInCoreData(gatewayResponse: result, authenticated: false) {
+        guard let phn = self.personalHealthNumber() else { return }
```
Comment (ConnorOgilvie91) [inline BCVaccineCard/BCVaccineCard/Screens/ReusableViewControllers/GatewayFormViewController.swift:565]: the id returned from func handleTestResultInCoreData is not the PHN, can you please revert this change

File: BCVaccineCard/BCVaccineCard/Screens/ReusableViewControllers/GatewayFormViewController.swift [hunk 2]
```diff
@@ -664,13 +669,12 @@ extension GatewayFormViewController: HealthGatewayAPIWorkerDelegate {
         }
     }
     
-    func handleTestResultInCoreData(gatewayResponse: GatewayTestResultResponse, authenticated: Bool) -> String? {
-        // Note, this first guard statement is to handle the case when health gateway is wonky - throws success with no error but has key nil values, so in this case we don't want to store a dummy patient value, as that's what was happening
-        guard let collectionDate = gatewayResponse.resourcePayload?.records.first?.collectionDateTime,
-              !collectionDate.trimWhiteSpacesAndNewLines.isEmpty, let reportID = gatewayResponse.resourcePayload?.records.first?.reportId,
-              !reportID.trimWhiteSpacesAndNewLines.isEmpty else { return nil }
+    func personalHealthNumber() -> String? {
```
Comment (ConnorOgilvie91) [inline BCVaccineCard/BCVaccineCard/Screens/ReusableViewControllers/GatewayFormViewController.swift:672]: These two aren't the same - the logic should not be stripped out of handleTestResultInCoreData

File: BCVaccineCard/BCVaccineCard/Screens/ReusableViewControllers/GatewayFormViewController.swift [hunk 3]
```diff
@@ -679,8 +683,16 @@ extension GatewayFormViewController: HealthGatewayAPIWorkerDelegate {
         } else {
             bday = nil
         }
-        guard let patient = StorageService.shared.fetchOrCreatePatient(phn: phn, name: gatewayResponse.resourcePayload?.records.first?.patientDisplayName, birthday: bday) else {return nil}
-        guard let object = StorageService.shared.storeTestResults(patient: patient ,gateWayResponse: gatewayResponse, authenticated: authenticated) else { return nil }
+        return bday
+    }
+    
+    func handleTestResultInCoreData(gatewayResponse: GatewayTestResultResponse, authenticated: Bool, phn: String, birthday: Date?) -> String? {
```
Comment (ConnorOgilvie91) [inline BCVaccineCard/BCVaccineCard/Screens/ReusableViewControllers/GatewayFormViewController.swift:689]: This looks like it's no longer being used - can you revert the changes on this file? These changes will cause the functionality to not work on prod

**Review by AmirShayegh — CHANGES_REQUESTED** [review]

File: BCVaccineCard/BCVaccineCard/CoreDataStack.swift [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+import Foundation
+import CoreData
+import EncryptedCoreData
+
+public protocol CoreDataStackProtocol {
+    var persistentContainer: NSPersistentContainer { get }
+}
+
+extension CoreDataStackProtocol {
```
[the remark this answers is not in the document]
Comment (AmirShayegh) [reply BCVaccineCard/BCVaccineCard/CoreDataStack.swift:16]: Also why is this a protocol and not a struct or class?

**Review by AmirShayegh — CHANGES_REQUESTED** [review]

File: BCVaccineCard/BCVaccineCard/CoreDataStack.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (freshfawzy) [reply BCVaccineCard/BCVaccineCard/CoreDataStack.swift:16]: just to have a class with single responsibility (CoreDataStack) which is responsible for creating the persistent container and managed object context used by StorageService

**Review by AmirShayegh — APPROVED** [review]

**Pull request merged**